python/ray/dag/compiled_dag_node.py

```diff
@@ -1075,6 +1192,202 @@ def _get_or_compile(
+            if not isinstance(task.dag_node, ClassMethodNode):
+                continue
+            for downstream_idx in task.downstream_node_idxs:
+                downstream_dag_node = self.idx_to_task[downstream_idx].dag_node
+                if isinstance(downstream_dag_node, MultiOutputNode):
+                    continue
+                graph[idx][DAGNodeOperationType.WRITE].add_edge(
+                    graph[downstream_idx][DAGNodeOperationType.READ]
+                )
+        return graph
+
+    def _build_execution_schedule(self):
+        """
+        Generate an execution schedule for each actor. The schedule is a list of
+        DAGNodeOperation.
+
+        Step 1: Generate a DAG node operation graph. Refer to the functions
+        `_generate_dag_operation_graph_node` and `_build_dag_node_operation_graph`
+        for more details.
+
+        Step 2: Topological sort
+
+        It is possible to have multiple DAGOperationGraphNodes with zero in-degree.
+        Refer to the function `_select_next_nodes` for the logic of selecting nodes.
+
+        Then, put the selected nodes into the corresponding actors' schedules.
+
+        The schedule should be intuitive to users, meaning that the execution should
+        perform operations in ascending order of `bind_index` as much as possible.
+
+        [Example]:
+
+        See `test_execution_schedule` for more examples.
+        """
+        # Step 1: Build a graph of DAGOperationGraphNode
+        actor_to_operation_nodes = self._generate_dag_operation_graph_node()
+        graph = self._build_dag_node_operation_graph(actor_to_operation_nodes)
+
+        actor_to_candidates: Dict[
+            "ray._raylet.ActorID", List[DAGOperationGraphNode]
+        ] = defaultdict(list)
+        for _, node_dict in graph.items():
+            for _, node in node_dict.items():
+                if node.in_degree == 0:
+                    heapq.heappush(
+                        actor_to_candidates[node.actor_handle._actor_id], node
+                    )
+
+        visited_nodes = set()
+
+        # Step 2: Topological sort
+        while actor_to_candidates:
+            nodes = _select_next_nodes(actor_to_candidates, graph)
+            for node in nodes:
+                if node in visited_nodes:
+                    continue
+                self.actor_to_execution_schedule[node.actor_handle].append(
+                    node.operation
+                )
+                visited_nodes.add(node)
+                for out_node_idx, out_node_type in node.out_edges:
+                    out_node = graph[out_node_idx][out_node_type]
+                    out_node.in_edges.remove((node.idx, node.operation.type))
+                    if out_node.in_degree == 0:
+                        heapq.heappush(
+                            actor_to_candidates[out_node.actor_handle._actor_id],
+                            out_node,
+                        )
+
+            delete_keys = []
```

Comment:
Why do we need to delete actors with empty queues? Actually, I'm surprised this works because it seems possible to have an actor's queue be temporarily empty (because of unfulfilled data dependencies). Please remove if it's not necessary and make sure the case I mentioned is unit-tested.

Reply:
Remove the empty queues so that when we iterate over `actor_to_candidates`, we don't encounter empty queues. In addition, `actor_to_candidates` is a `defaultdict(list)`